doc/developer/design/20230110_window_functions.md

```diff
@@ -0,0 +1,255 @@
+
+There are several options for how to represent windowing in HIR, MIR, and LIR. For each of the IRs, I can see 3 options:
+
+1. Create a new relation expression enum variant. This could be a dedicated variant just for windowing, or it could be a many-to-many Reduce, which would initially only handle windowing, but later we could also merge `TopK` into it. (Standard Reduce is N-to-1, TopK is N-to-K, a window function is N-to-N. There are differences also in output columns.)
+2. Hide away windowing stuff in scalar expressions. (the current way in HIR)
+3. Reuse an existing relation expression enum variant, e.g., `Reduce`.
+
+I propose doing 1. for MIR and LIR, and later have a separate EPIC to consider unifying windowing and Reduce (and potentially TopK) into a many-to-many “super-reduce”. (We might only unify it in MIR, but not LIR. But we can decide this later.)
+
+There is a discussion below on each of the three above options for each of the IRs, but first we discuss another option, which bypasses any MIR/LIR modifications.
+
+### Recognizing the current windowing pattern in just the rendering
+
+There is one more option that bypasses any MIR/LIR modifications: we could leave HIR, MIR, and LIR unchanged, and just make the rendering detect that pattern that the current HIR lowering creates, and do a custom rendering to it. However, this might be too brittle: we might accidentally leave out cases when the pattern somehow ends up slightly different, plus we might break it from time to time with unrelated changes. An example, for what gets created by the current HIR lowering from a window function call (`LAG`):
+
+```c
+materialize=> explain select name, pop, LAG(name) OVER (partition by state order by pop)
+from cities;
+                                          Optimized Plan
+--------------------------------------------------------------------------------------------------
+ Explained Query:                                                                                +
+   Project (#3..=#5)                                                                             +
+     Map (record_get[1](#1), record_get[0](#2), record_get[2](#2), record_get[0](#1))            +
+       FlatMap unnest_list(#0)                                                                   +
+         Project (#1)                                                                            +
+           Reduce group_by=[#1] aggregates=[lag(row(row(row(#0, #1, #2), row(#0, 1, null)), #2))]+
+             Get materialize.public.cities                                                       +
+```
+
+(And this is just the MIR, but we would work with the LIR.)
+
+Also, we will want other transforms for window functions, most notably, the ROW_NUMBER-to-TopK transform. For this, we need to canonicalize scalar expressions, which I think we usually do in MIR. This means that this transform should happen on MIR, meaning that we would need an MIR representation of window functions.
```

Comment:
I struggle to understand your use of "need" here. It seems not hard to write a transform that looks for the expected idiom of a filter over the `row_number` window function block. That transform may not be great, it may be brittle, but it is a thing we could do. 

Is there a technical **requirement** to have an MIR variant for window functions, or is it just something that will make window functions themselves easier to implement and maintain (at the expense of the rest of the codebase)?

Reply:
> Is there a technical **requirement** to have an MIR variant for window functions

Ok, there is no requirement. I changed the design doc to go with the idiom recognition approach also for the TopK pattern.